blog/2025-12-signal-forms-part4/README.md

```diff
@@ -0,0 +1,336 @@
+
+The directive receives the field as an input.
+It also accepts another input with the ID of the HTML element that contains the related messages (info, errors, loading).
+We will need this ID for the attributes `aria-describedby` or `aria-errormessage`, to connect the field with the related messages.
+We link the id with `aria-errormessage` when the field is currently invalid and touched.
+Otherwise we link it with `aria-describedby`.
+
+To write to the ARIA attributes, we use Host Bindings in the directive's metadata.
+
+```typescript
+// field-aria-attributes.ts
+import { computed, Directive, input } from '@angular/core';
+import { FieldTree } from '@angular/forms/signals';
+
+@Directive({
+  selector: '[field]',
+  host: {
+    '[aria-invalid]': 'ariaInvalid()',
+    '[aria-busy]': 'ariaBusy()',
+    '[aria-describedby]': 'ariaDescribedBy()',
+    '[aria-errormessage]': 'ariaErrorMessage()',
+  },
+})
+export class FieldAriaAttributes<T> {
+  readonly field = input.required<FieldTree<T>>();
+  readonly fieldDescriptionId = input<string>();
+
+  readonly ariaInvalid = computed(() => {
+    const state = this.field()();
+    return state.touched() && !state.pending() ? state.errors().length > 0 : undefined;
+  });
+
+  readonly ariaBusy = computed(() => {
+    const state = this.field()();
+    return state.pending();
+  });
+
+  readonly ariaDescribedBy = computed(() => {
+    const id = this.fieldDescriptionId();
+    return !id || this.ariaInvalid() ? null : id;
+  });
+
+  readonly ariaErrorMessage = computed(() => {
+    const id = this.fieldDescriptionId();
+    return !id || !this.ariaInvalid() ? null : id;
+  });
+}
+```
+
+The directive now manages four key ARIA attributes:
+
+- **`aria-invalid`**: set to `true` when the field has been touched and contains validation errors.
+- **`aria-busy`**: set to `true` during async validation to indicate loading state.
+- **`aria-describedby`**: references the ID of the element containing helpful information (when field is valid).
+- **`aria-errormessage`**: references the ID of the element containing error messages (when field is invalid).
+
+### Using the Directive
+
+The directive works seamlessly with the existing `Field` directive since we defined the selector to be `[field]`.
+To use it, we need to import it in our component holding the form in the component's decorator.
+
+```typescript
+@Component({
+  selector: 'app-registration-form',
+  imports: [Field, FormFieldInfo, FieldAriaAttributes, /* other imports */],
+  // ...
+})
+export class RegistrationForm {
+  // ...
+}
+```
+
+Basically that's it: The attributes `aria-invalid` and `aria-busy` are automatically set since our directive binds to the existing `[field]` directive.
+To make it really accessible we need to pass the ID of the targeting message element as the `fieldDescriptionId` input to connect the field with its description element:
+
```

Comment:
Wir hatten glaube in den Iterationen vorher noch die Attribute manuell gesagt und diese hilfsmethode in der Komponete. Hier sollte noch mal der Hinweis irgendwo kommen, dass wir das nicht mehr brauchen.